include/singa/io/encoder.h

```diff
@@ -30,13 +31,15 @@ class Encoder {
   public:
     Encoder() { }
     virtual ~Encoder() { }
+    
+    virtual void Setup(const LayerConf& conf) = 0;
```

Comment:
do not make such assumption.
encoders and decoders should be independent from layer.